src/adaptix/_internal/model_tools/introspection/msgspec.py

```diff
@@ -0,0 +1,108 @@
+import inspect
+from types import MappingProxyType
+from typing import Mapping
+
+try:
+    from msgspec import NODEFAULT
+    from msgspec.structs import FieldInfo, fields
+except ImportError:
+    pass
+from ...feature_requirement import HAS_MSGSPEC_PKG, HAS_SUPPORTED_MSGSPEC_PKG
+from ...type_tools import get_all_type_hints, is_class_var, normalize_type
+from ..definitions import (
+    Default,
+    DefaultFactory,
+    DefaultValue,
+    FullShape,
+    InputField,
+    InputShape,
+    IntrospectionError,
+    NoDefault,
+    NoTargetPackageError,
+    OutputField,
+    OutputShape,
+    Param,
+    ParamKind,
+    TooOldPackageError,
+    create_attr_accessor,
+)
+
+
+def _get_default_from_field_info(fi: "FieldInfo") -> Default:
+    if fi.default is not NODEFAULT:
+        return DefaultValue(fi.default)
+    if fi.default_factory is not NODEFAULT:
+        return DefaultFactory(fi.default_factory)
+    return NoDefault()
+
+
+def _create_input_field_from_structs_field_info(fi: "FieldInfo", type_hints: Mapping[str, type]) -> InputField:
+    default = _get_default_from_field_info(fi)
+    return InputField(
+        id=fi.name,
+        type=type_hints[fi.name],
+        default=default,
+        is_required=fi.required,
+        original=fi,
+        metadata=MappingProxyType({}),
+    )
+
+
+def get_struct_shape(tp) -> FullShape:
+    if not HAS_SUPPORTED_MSGSPEC_PKG:
+        if not HAS_MSGSPEC_PKG:
+            raise NoTargetPackageError(HAS_MSGSPEC_PKG)
+        raise TooOldPackageError(HAS_SUPPORTED_MSGSPEC_PKG)
+
+    try:
+        fields_info = fields(tp)
+    except TypeError:
+        raise IntrospectionError
+
+    type_hints = get_all_type_hints(tp)
+    init_fields = tuple(
+        field_name
+        for field_name in type_hints
+        if not is_class_var(normalize_type(type_hints[field_name]))
+    )
+    input_param_kind = ParamKind.KW_ONLY if "*" in inspect.signature(tp) else ParamKind.POS_OR_KW
```

Comment:
This looks very unreliable. You can get `ParameterKind` of each parameter from signature object and translate it to `ParamrKind`